### ISSUE TITLE
Don't use a mutex for computing Reference hash

### ISSUE THREAD

**Issue by pepone** [zeroc-ice/ice#1709]
I don't think we should use any mutex in this code.

Either the hash is very cheap to compute and we compute it each time. Or it's somewhat expensive to compute (== bad hash) and we compute it in the constructor. As far as I can tell, this class is immutable.

_Originally posted by @bernardnormier in https://github.com/zeroc-ice/ice/pull/1707#pullrequestreview-1836571809_

**Comment by bernardnormier** [comment]
Fixed by #2120 in C#. 
Fixed by #2123 in C++.

**Comment by InsertCreativityHere** [comment]
Fixed by #2437 in Java.
And Javascript has no mutex locking.

**Comment by bernardnormier** [comment]
```
ice % rg hashInit
js/src/Ice/Reference.js
45:        this._hashInitialized = false;
265:        if (this._hashInitialized) {
285:        this._hashInitialized = true;
990:        if (!this._hashInitialized) {

js/src/Ice/IPEndpointI.js
77:            this._hashCode = this.hashInit(5381);
170:    hashInit(h) {

js/src/Ice/TcpEndpointI.js
185:    hashInit(h) {
186:        h = super.hashInit(h);
```

Unfortunately, while JS doesn't use a mutex, it uses the undesirable "hash value caching" logic. Still needs to be fixed.

**Comment by pepone** [comment]
JavaScript fixed in #2607
